src/tracks/playabletrack/wavetrack/ui/SpectrumView.h

```diff
@@ -11,32 +11,120 @@ Paul Licameli split from WaveTrackView.h
 #ifndef __AUDACITY_SPECTRUM_VIEW__
 #define __AUDACITY_SPECTRUM_VIEW__
 
+#include <map>
+#include <set>
 #include "WaveTrackView.h" // to inherit
 
+
 class WaveTrack;
+class BrushHandle;
+using TimeFreqBinsMap = std::map<long long, std::set<wxInt64>>;
+
+class SpectralData{
+private:
+   double mSampleRate;
+   long long mStartT = std::numeric_limits<long long>::max();
+   long long mEndT = 0;
+
+public:
+    SpectralData(double sr)
+    :mSampleRate(sr)
+    ,mStartT(std::numeric_limits<long long>::max())
+    ,mEndT( 0 )
+    {}
+    SpectralData(const SpectralData& src) = delete;
+
+   TimeFreqBinsMap dataBuffer;
+   std::vector<TimeFreqBinsMap> dataHistory;
+   // TODO: replace with two pairs to save space
+   std::vector<std::pair<int, int>> coordHistory;
+
+   double GetSR() const{
+      return mSampleRate;
+   }
+
+   long long GetStartT() const{
+      return mStartT;
+   }
+
+   long long GetEndT() const{
+      return mEndT;
+   }
+
+   // The double time points is quantized into long long
+   void addTimeFreqData(long long ll_sc, wxInt64 freq){
+      if(ll_sc > mEndT)
+         mEndT = ll_sc;
+
+      if(ll_sc < mStartT)
+         mStartT = ll_sc;
+
+      if(dataBuffer.find(ll_sc) == dataBuffer.end())
+         dataBuffer[ll_sc] = std::set<wxInt64>{ freq };
+      else
+         dataBuffer[ll_sc].insert(freq);
+   }
+
+   void removeTimeFreqData(long long ll_sc, wxInt64 freq){
+      for(auto &dataBuf: dataHistory){
+         if(dataBuf.find(ll_sc) != dataBuf.end()){
+            dataBuf[ll_sc].erase(freq);
+         }
+      }
+   }
```

Comment:
I'm puzzled why the add function affects only the current state, but the subtract function rewrites all past states.

Maybe if you rounded times to discrete hops of 512 (or whatever window size / 4 is), as I suggested elsewhere, then you would find it easier to erase what you added just in the most recent state.

Reply:
I think I understand things better now, and it's the right thing to do after all.